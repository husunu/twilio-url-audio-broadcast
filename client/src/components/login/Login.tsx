import React, { useState, useCallback, useMemo, useEffect } from 'react';
import useFetch from 'use-http';
import { useHistory, useLocation } from 'react-router';
import { Typography, Grid, Box, TextField, Button } from '@material-ui/core';
import TermsAndConditions from './TermsAndConditions';
import useStyles from './styles';

import {
  FEEDBACK_ERROR_MSG,
  LOGIN_BUTTON_TEST_ID,
  PASSCODE_INPUT_TEST_ID,
  REDIRECT_URL,
  SESSION_STORAGE_PASSCODE_KEY,
} from './constants';

export const Login = () => {
  const classes = useStyles();
  const history = useHistory();
  const [request, response] = useFetch();
  const [passcode, setPasscode] = useState('');
  const [hasFailed, setHasFailed] = useState(false);
  const location = useLocation<{ from: string }>();

  const clearError = useCallback(() => {
    if (hasFailed) setHasFailed(false);
  }, [hasFailed, setHasFailed]);

  const isDisabled = useMemo(() => !passcode.trim().length || request.loading, [
    passcode,
    request.loading,
  ]);

  const onChangePassCode = useCallback(
    (event: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => {
      clearError();
      setPasscode(event.target.value);
    },
    [setPasscode, clearError]
  );

  const onSubmit = useCallback(
    async (event: React.FormEvent<HTMLFormElement>) => {
      event.preventDefault();
      clearError();
      await request.post('/login', { passcode });
      if (response.ok) {
        sessionStorage.setItem(SESSION_STORAGE_PASSCODE_KEY, passcode);
        history.push(location?.state?.from || REDIRECT_URL);
      }
    },
    [clearError, request, passcode, response.ok, history, location]
  );

  useEffect(() => {
    setHasFailed(Boolean(request.error));
  }, [request.error]);

  useEffect(() => {
    const params = new URLSearchParams(location.search);
    const passcodeParam = params.get('passcode');

    if (passcodeParam) {
      setPasscode(passcodeParam);
    }
  }, [location, setPasscode]);

  return (
    <>
      <Typography variant="h4" component="h1" gutterBottom={true}>
        Send notifications to your customers over the phone
      </Typography>
      <Typography>
        In order to use the app you need to authenticate. Please login using the
        passcode generated during installation and app deployment
      </Typography>
      <Grid container spacing={3}>
        <Grid item xs={8}>
          <form
            onSubmit={onSubmit}
            className={classes.root}
            noValidate
            autoComplete="off"
          >
            <TextField
              id="passcode"
              label="Enter your passcode"
              variant="outlined"
              size="small"
              onChange={onChangePassCode}
              value={passcode}
              error={hasFailed}
              type="password"
              helperText={hasFailed && FEEDBACK_ERROR_MSG}
              InputProps={{
                inputProps: {
                  'data-testid': PASSCODE_INPUT_TEST_ID,
                },
              }}
            />
            <Box mt={4}>
              <TermsAndConditions />
            </Box>
            <Box mt={4}>
              <Button
                variant="contained"
                color="primary"
                disabled={isDisabled}
                data-testid={LOGIN_BUTTON_TEST_ID}
                type="submit"
              >
                Login
              </Button>
            </Box>
          </form>
        </Grid>
      </Grid>
    </>
  );
};
