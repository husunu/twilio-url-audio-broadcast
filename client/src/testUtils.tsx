import React, { ReactNode } from 'react';
import { Router } from 'react-router';
import { render } from '@testing-library/react';
import { createMemoryHistory, MemoryHistory } from 'history';
import { Provider } from 'react-redux';

import { store } from '@/redux';
import { HttpProvider } from '@/http-provider';

const makeContextWrapper = (
  history: MemoryHistory
): React.FC<{
  children?: ReactNode;
}> => ({ children }) => (
  <Provider store={store}>
    <Router history={history}>
      <HttpProvider>{children}</HttpProvider>
    </Router>
  </Provider>
);

export const renderWithAppContexts = (
  ui: React.ReactElement,
  options?: any
) => {
  const history = createMemoryHistory();
  const Wrapper = makeContextWrapper(history);
  return {
    ...render(ui, { wrapper: Wrapper, ...options }),
    history,
  };
};
