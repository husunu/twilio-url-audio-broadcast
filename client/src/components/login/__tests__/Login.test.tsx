import React from 'react';
import { Route } from 'react-router-dom';
import { fireEvent, waitFor, cleanup } from '@testing-library/react';
import { FetchMock } from 'jest-fetch-mock';
import { renderWithAppContexts } from '@/testUtils';
import { Login } from '../Login';

import {
  FEEDBACK_ERROR_MSG,
  LOGIN_BUTTON_TEST_ID,
  PASSCODE_INPUT_TEST_ID,
  REDIRECT_URL,
  SESSION_STORAGE_PASSCODE_KEY,
} from '@/components/login/constants';

const fetch = global.fetch as FetchMock;

describe('Login', () => {
  const samplePasscode = '0123456789';

  const fillPasscode = (getByTestId: Function, passcode: string) => {
    fireEvent.change(getByTestId(PASSCODE_INPUT_TEST_ID), {
      target: { value: passcode },
    });
  };

  const fillAndSubmitLogin = (getByTestId: Function) => {
    fillPasscode(getByTestId, samplePasscode);
    fireEvent.click(getByTestId(LOGIN_BUTTON_TEST_ID), { button: 0 });
  };

  afterEach(() => {
    cleanup();
    fetch.resetMocks();
  });

  test('renders the page', () => {
    const { container } = renderWithAppContexts(<Login />);
    expect(container).toMatchSnapshot();
  });

  test('when login is successful user is redirected to broadcast list', async () => {
    fetch.mockResponse(JSON.stringify({}), { status: 204 });

    const dummyTestId = 'dummyTestId';
    const { getByTestId, history } = renderWithAppContexts(
      <>
        <Route path="/">
          <Login />
        </Route>
        <Route path={REDIRECT_URL}>
          <span data-testid={dummyTestId} />
        </Route>
      </>
    );

    fillAndSubmitLogin(getByTestId);

    await waitFor(() => {
      expect(getByTestId(dummyTestId)).toBeInTheDocument();
      expect(history.location.pathname).toEqual(REDIRECT_URL);
      expect(sessionStorage.getItem(SESSION_STORAGE_PASSCODE_KEY)).toEqual(
        samplePasscode
      );
    });
  });

  test('when login is unsuccessful a message error is shown and removed when passcode changes', async () => {
    fetch.mockResponse(JSON.stringify(new Error('Bad Request')), {
      status: 400,
    });
    const { getByText, getByTestId, queryByText } = renderWithAppContexts(
      <Login />
    );

    fillAndSubmitLogin(getByTestId);

    await waitFor(() => {
      expect(getByText(FEEDBACK_ERROR_MSG)).toBeInTheDocument();
      fillPasscode(getByTestId, '1234');
      expect(queryByText(FEEDBACK_ERROR_MSG)).not.toBeInTheDocument();
    });
  });

  test('login button is disabled when passcode is empty and enabled when it is filled', () => {
    const { getByTestId } = renderWithAppContexts(<Login />);
    const button = getByTestId(LOGIN_BUTTON_TEST_ID);
    expect(button).toBeDisabled();
    fillPasscode(getByTestId, samplePasscode);
    expect(button).not.toBeDisabled();
    fillPasscode(getByTestId, '');
    expect(button).toBeDisabled();
  });

  test('login button is disabled when login request is loading', async () => {
    fetch.mockResponse(
      () =>
        new Promise((resolve, reject) =>
          setTimeout(() => reject({ body: 'ko' }), 100)
        )
    );
    const { getByTestId } = renderWithAppContexts(<Login />);
    const button = getByTestId(LOGIN_BUTTON_TEST_ID);
    fillAndSubmitLogin(getByTestId);

    await waitFor(() => {
      expect(button).toBeDisabled();
    });
    await waitFor(() => {
      expect(button).not.toBeDisabled();
    });
  });
});
