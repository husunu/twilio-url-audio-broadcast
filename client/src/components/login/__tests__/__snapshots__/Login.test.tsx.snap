// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login renders the page 1`] = `
<div>
  <h1
    class="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom"
  >
    Send notifications to your customers over the phone
  </h1>
  <p
    class="MuiTypography-root MuiTypography-body1"
  >
    In order to use the app you need to authenticate. Please login using the passcode generated during installation and app deployment
  </p>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-8"
    >
      <form
        autocomplete="off"
        class="makeStyles-root-1"
        novalidate=""
      >
        <div
          class="MuiFormControl-root MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense MuiInputLabel-outlined"
            data-shrink="false"
            for="passcode"
            id="passcode-label"
          >
            Enter your passcode
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-marginDense MuiOutlinedInput-marginDense"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense"
              data-testid="passcode-input"
              id="passcode"
              type="password"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="PrivateNotchedOutline-root-194 MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="PrivateNotchedOutline-legendLabelled-196"
              >
                <span>
                  Enter your passcode
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="MuiBox-root MuiBox-root-199"
        >
          <div
            class="MuiPaper-root MuiAlert-root MuiAlert-standardInfo MuiPaper-elevation0"
            role="alert"
          >
            <div
              class="MuiAlert-icon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20, 12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10, 10 0 0,0 12,2M11,17H13V11H11V17Z"
                />
              </svg>
            </div>
            <div
              class="MuiAlert-message"
            >
              <strong>
                Reminder 
              </strong>
              - Please note you have agreed to the
               
              <a
                class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
                href="https://www.twilio.com/legal/tos"
              >
                Twilio Terms and Service
              </a>
               
              and
               
              <a
                class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
                href="https://www.twilio.com/legal/aup"
              >
                Twilio Acceptable Use Policy
              </a>
              , as incorporated therein.
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root MuiBox-root-259"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary Mui-disabled Mui-disabled"
            data-testid="login-button"
            disabled=""
            tabindex="-1"
            type="submit"
          >
            <span
              class="MuiButton-label"
            >
              Login
            </span>
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
